ray-on-gke/kuberay-tpu-webhook/main.go

```diff
@@ -123,6 +124,84 @@ func injectHostnames(hostNames string, workerGroupSpec ray.WorkerGroupSpec, work
 	}
 }
 
+func checkWorkersMatchTopology(workerGroupSpec ray.WorkerGroupSpec) (bool, error) {
+	numWorkers := int(*workerGroupSpec.Replicas)
+	if containerRequestingTPUs(workerGroupSpec.Template.Spec.Containers...) {
+		topology := workerGroupSpec.Template.Spec.NodeSelector["cloud.google.com/gke-tpu-topology"]
+		hosts, err := getNumTPUHosts(topology)
+		if err != nil {
+			return false, err
+		}
+
+		if hosts != numWorkers {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func checkWorkerGroupAffinity(workerGroupSpec ray.WorkerGroupSpec) (bool, error) {
+	if containerRequestingTPUs(workerGroupSpec.Template.Spec.Containers...) {
+		topology := workerGroupSpec.Template.Spec.NodeSelector["cloud.google.com/gke-tpu-topology"]
+		isMultiHost, err := isTPUMultiHost(topology)
+		if err != nil {
+			return false, err
+		}
+		if isMultiHost {
+			placementGroup := workerGroupSpec.Template.Spec.NodeSelector["cloud.google.com/gke-placement-group"]
+			if placementGroup == "" {
+				return false, nil	
+			}
+		}
+	}
+
+	return true, nil
+}
+
+func validateRayCluster(admissionReview *admissionv1.AdmissionReview) (*admissionv1.AdmissionResponse, error) {
+	raycluster, err := extractRayCluster(admissionReview)
+	if err != nil {
+		klog.Fatalf("Ray Cluster extraction failed: %s", err)
+	}
+
+	admit := true
+	status := "Success"
+	message := ""
+	for i := 0; i < len(raycluster.Spec.WorkerGroupSpecs); i++ {
+		workerGroupSpec := raycluster.Spec.WorkerGroupSpecs[i]
+		workersMatchTopology, err := checkWorkersMatchTopology(workerGroupSpec)
+		if err != nil {
+			return nil, err
+		}
+		workerGroupAffinity, err := checkWorkerGroupAffinity(workerGroupSpec)
+		if err != nil {
+			return nil, err
+		}
+
+		if !(workersMatchTopology && workerGroupAffinity) {
+			admit = false
+			status = "Failure"
+			if !workersMatchTopology {
```

Comment:
If both of these errors can occur at the same time (workers don't match topology, worker group affinity set incorrectly, should the error message show both of these, instead of just one?

Reply:
Done.